### PR TITLE
The Mailer Documentation

### PR DESCRIPTION
**Pull request by weaverryan**
o/ friends!

Say hello to the Mailer documentation (#11626 does the component).

A lot of the docs were moved from the mime component, and pretty much everything here was tested. Reviews, questions and comments are warmly welcomed.

Note: the first commit was just to "deprecate" the Swift Mailer docs and copy the `email/*` docs into the one, deprecated Swift Mailer document (`email.rst`). Mailer lives in `mailer.rst`.

Cheers!

### CODE REVIEW

File: _build/redirection_map [hunk 1]
```diff
@@ -417,3 +417,6 @@
 /workflow/state-machines /workflow/introduction
 /workflow/usage /workflow
 /introduction/from_flat_php_to_symfony2 /introduction/from_flat_php_to_symfony
+/email/dev_environment /mailer
+/email/spool /mailer
+/email/testing /mailer
```
Comment (OskarStark) [inline _build/redirection_map]: please add a newline

File: email.rst [hunk 2]
```diff
@@ -1,8 +1,12 @@
 .. index::
    single: Emails
 
-How to Send an Email
-====================
+Swift Mailer
+============
+
+.. note::
```
Comment (OskarStark) [inline email.rst]: please add a blank line afterwards

**Review by OskarStark — APPROVED** [review]

File: mailer.rst [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+Twig: HTML & CSS
+----------------
+
+The Mime component integrates with the :doc:`Twig template engine </templating>`
+to provide advanced features such as CSS style inlining and support for HTML/CSS
+frameworks to create complex HTML email messages. First, make sure Twig is instaled:
+
+.. code-block:: terminal
+
+    $ composer require twig
```
Comment (wouterj) [inline mailer.rst]:
```suggestion
    $ composer require symfony/twig-bundle
```

IIRC, we decided not to use aliases in the docs. (and you're not using it anywhere else in this guide)

File: mailer.rst [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+
+    $email = (new TemplatedEmail())
+        ->from('fabien@symfony.com')
+        ->to(new NamedAddress('ryan@symfonycasts.com', 'Ryan'))
+        ->subject('Thanks for signing up!')
+
+        // path of the Twig template to render
+        ->htmlTemplate('emails/signup.html.twig')
+
+        // pass variable (name => value) to the template
```
Comment (wouterj) [inline mailer.rst]:
```suggestion
        // pass variables (name => value) to the template
```

File: mailer.rst [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+        You signed up as {{ username }} the following email:
+    </p>
+    <p><code>{{ email.to[0].address }}</code></p>
+
+    <p>
+        <a href="#">Click here to activate your account</a>
+        (this link is valid until {{ expiration_date|date('F jS') }})
+    </p>
+
+
```
Comment (wouterj) [inline mailer.rst]: double empty line

File: mailer.rst [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+    + use Symfony\Bridge\Twig\Mime\TemplatedEmail;
+
+    $email = (new TemplatedEmail())
+        // ...
+
+        ->htmlTemplate('emails/signup.html.twig')
+    +     ->textTemplate('emails/signup.txt.twig')
+        // ...
+    ;
```
Comment (wouterj) [inline mailer.rst]: I believe correct diff templates should indent all lines with 2 spaces, except from the ones containing `+ `/`- `)

File: mailer.rst [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+
+Using External CSS Files
+........................
+
+You can also define CSS styles in external files and pass them as
+arguments to the filter:
+
+.. code-block:: html+twig
+
+    {# '@css/' refers to the Twig namespace defined earlier #}
```
Comment (wouterj) [inline mailer.rst]: it's not defined earlier?

File: mailer.rst [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+    {% filter inline_css(source('@css/email.css')) %}
+        <style>
+            {# here, define your CSS styles as usual #}
```
Comment (wouterj) [inline mailer.rst]: not so sure about this, isn't the idea of this example to show *not* using `<style>` but rather using a CSS file?

File: mailer.rst [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+
+    # config/packages/twig.yaml
+    twig:
+        # ...
+
+        paths:
+            # point this anywhere that your css files live
+            css: '%kernel.project_dir%/assets/css'
+
+You can pass unlimited number of arguments to ``inline_css()`` to load multiple
```
Comment (wouterj) [inline mailer.rst]: To me, this belongs better directly after the Twig template

File: mailer.rst [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+
+You can combine all filters to create complex email messages:
+
+.. code-block:: twig
+
+    {% filter inky|inline_css(source('@css/foundation-emails.css')) %}
+        {# ... #}
+    {% endfilter %}
+
+This makes use of the :ref:`css Twig namespace<mailer-css-namespace>` we created
```
Comment (wouterj) [inline mailer.rst]:
```suggestion
This makes use of the :ref:`css Twig namespace <mailer-css-namespace>` we created
```

**Review by wouterj — CHANGES_REQUESTED** [review]
Very nice job! And I 😍 PHP 7+ code examples

The section talking about all twig filters (inky, etc) seems to me like it should be belong to a sub-guide (with a small reference section in the main guide that briefly mentions all integratable tools). You're often interested in just one and reading every single one seems a bit too much. Other than that major +1

File: mailer.rst [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+==========================
+
+.. versionadded:: 4.3
+    The Mailer component was added in Symfony 4.3. Previously, Swift Mailer
+    was recommended. You can still :doc:`read its documentation</email/swiftmailer>`.
+
+Installation
+------------
+
+Symfony's Mailer component is a *powerful* way to send email - complete with
```
Comment (fabpot) [inline mailer.rst]: Here, you're talking about the Mailer AND the MIME component, not just the Mailer component. The difference is huge and important. Mailer is *only* about sending emails. MIME is about creating emails. Having one document to explain how to create and send emails is ok, but I think we need to keep "deep" docs in the component docs themselves.

File: mailer.rst [hunk 12]
```diff
@@ -0,0 +1,653 @@
...
+Gmail               ``composer require symfony/google-mailer``
+==================  =============================================
+
+Each library includes a :ref:`Flex recipe <flex-recipe>` that will add example configuration
+to your ``.env`` file. For example, suppose you want to use SendGrid. First,
+install it:
+
+.. code-block:: terminal
+
+    $ composer require symfony/symfony/sendgrid-mailer
```
Comment (harikt) [inline mailer.rst]: `symfony/symfony/sendgrid-mailer` , only one symfony right ?

File: mailer.rst [hunk 13]
```diff
@@ -0,0 +1,653 @@
...
+All the methods that require email addresses (``from()``, ``to()``, etc.) accept
+both strings and objects::
+
+    // ...
+    use Symfony\Component\Mime\Address;
+    use Symfony\Component\Mime\NamedAddress;
+
+    $email = (new Email())
+        // email address as a simple string
+        ->from('fabien@symfony.com')
```
Comment (harikt) [inline mailer.rst]: Introducing a real email will spam people. It may be good to use example.com instead of real domain.

File: mailer.rst (hunk 12 above)
Comment (ricknox) [reply mailer.rst]:
```suggestion
    $ composer require symfony/sendgrid-mailer
```

File: mailer.rst [hunk 14]
```diff
@@ -0,0 +1,653 @@
...
+Amazon SES          ``composer require symfony/amazon-mailer``
+MailChimp           ``composer require symfony/mailchimp-mailer``
+Mailgun             ``composer require symfony/mailgun-mailer``
+Postmark            ``composer require symfony/postmark-mailer``
+SendGrid            ``composer require symfony/sendgrid-mailer``
+Gmail               ``composer require symfony/google-mailer``
+==================  =============================================
+
+Each library includes a :ref:`Flex recipe <flex-recipe>` that will add example configuration
+to your ``.env`` file. For example, suppose you want to use SendGrid. First,
```
Comment (ricknox) [inline mailer.rst]:
```suggestion
to the ``.env`` file. For example, suppose you want to use SendGrid. First,
```

File: mailer.rst (hunk 6 above)
Comment (weaverryan) [reply mailer.rst]: I think what I have is correct. Looking at other parts of the docs, a diff code block (at least the way I've been creating them) involves making a normal code block (with normal indentation) and then adding `+ ` or `- ` (plus then space) on the changed lines. That indeed makes those lines look a little "funny" - as the changed lines don't line up with the unchanged lines inside the raw rst (but it renders correctly).

File: email.rst [hunk 15]
```diff
@@ -158,16 +163,498 @@ file. For example, for `Amazon SES`_ (Simple Email Service):
...
+    .. code-block:: php
+
+        // config/packages/dev/web_profiler.php
+        $container->loadFromExtension('web_profiler', [
+            'intercept_redirects' => 'true',
+        ]);
+
+.. tip::
+
+    Alternatively, you can open the profiler after the redirect and search
```
Comment (noniagriconomie) [inline email.rst]: what about adding this directly after the previous paragraph?
not as a tip, but as a two way options

personnaly i have never used intercept_redirects

File: mailer.rst [hunk 16]
```diff
@@ -0,0 +1,649 @@
...
+
+.. code-block:: terminal
+
+    $ composer require symfony/mailer
+
+Transport Setup
+---------------
+
+Emails are delivered via a "transport". And without installing anything else, you
+delivery emails over ``smtp`` by configuring your ``.env`` file:
```
Comment (noniagriconomie) [inline mailer.rst]: sentence seems wrong when reading, is a word missing?

File: mailer.rst [hunk 17]
```diff
@@ -0,0 +1,649 @@
...
+Using a 3rd Party Transport
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+But an easier option is to send emails via a 3rd party provider. Mailer supports
+several - install whichever you want:
+
+==================  =============================================
+Service             Install with
+==================  =============================================
+Amazon SES          ``composer require symfony/amazon-mailer``
```
Comment (noniagriconomie) [inline mailer.rst]: the order is by popularity or random?

File: mailer.rst [hunk 18]
```diff
@@ -0,0 +1,649 @@
...
+        public function sendEmail(MailerInterface $mailer)
+        {
+            $email = (new Email())
+                ->from('hello@example.com')
+                ->to('you@example.com')
+                //->cc('cc@example.com')
+                //->bcc('bcc@example.com')
+                //->replyTo('fabien@example.com')
+                //->priority(Email::PRIORITY_HIGH)
+                ->subject('Time for Symfony Mailer!!!99')
```
Comment (noniagriconomie) [inline mailer.rst]: what is the 99 ?

File: mailer.rst [hunk 19]
```diff
@@ -0,0 +1,649 @@
...
+        }
+    }
+
+That's it! The message will be sent via whatever transport you configured.
+
+Email Addresses
+~~~~~~~~~~~~~~~
+
+All the methods that require email addresses (``from()``, ``to()``, etc.) accept
+both strings and objects::
```
Comment (noniagriconomie) [inline mailer.rst]:
```suggestion
both strings and `Mime` objects::
```

File: mailer.rst [hunk 20]
```diff
@@ -0,0 +1,649 @@
...
+        ->html('<img src="cid:logo"> ... <img src="cid:footer-signature"> ...')
+    ;
+
+Global from Address
+-------------------
+
+Instead of calling ``->from()`` *every* time you create a new email, you can
+create an event subscriber to set it automatically::
+
+    // src/Mailer/MailerFromListener.php
```
Comment (noniagriconomie) [inline mailer.rst]: better inside an EventListener folder?

File: mailer.rst [hunk 21]
```diff
@@ -0,0 +1,649 @@
...
+method of the ``TemplatedEmail`` class and also to a special variable called
+``email``, which is an instance of
+:class:`Symfony\\Bridge\\Twig\\Mime\\WrappedTemplatedEmail`.
+
+Text Content
+~~~~~~~~~~~~
+
+When the text content of a ``TemplatedEmail`` is not explicitly defined, mailer
+will generate it automatically by converting the HTML contents into text. If you
+have `league/html-to-markdown`_ installed in your application,
```
Comment (noniagriconomie) [inline mailer.rst]: cool one

File: mailer.rst [hunk 22]
```diff
@@ -0,0 +1,649 @@
...
+``images`` that points to whatever directory your images are stored in:
+
+.. code-block:: yaml
+
+    # config/packages/twig.yaml
+    twig:
+        # ...
+
+        paths:
+            # point this anywhere that your images live
```
Comment (noniagriconomie) [inline mailer.rst]: anywhere that, feel strange reading

**Review by noniagriconomie — COMMENTED** [review]
nice add, wanted to learn about it, PR doc is a good starting point :)

File: email.rst [hunk 23]
```diff
@@ -158,16 +163,498 @@ file. For example, for `Amazon SES`_ (Simple Email Service):
...
+    .. code-block:: php
+
+        // config/packages/dev/swiftmailer.php
+        $container->loadFromExtension('swiftmailer', [
+            'delivery_addresses' => ['dev@example.com'],
+        ]);
+
+Now, suppose you're sending an email to ``recipient@example.com`` in a controller::
+
+    public function index($name, \Swift_Mailer $mailer)
```
[the remark this answers is not in the document]
Comment (weaverryan) [reply email.rst]: You're right, but this is just moved from another spot in the docs - we won't update Swiftmailer docs for php7 - just not worth it (well, at least not in this PR)

File: email.rst (hunk 15 above)
Comment (weaverryan) [reply email.rst]: good note, but can you propose in a separate PR? This is just moved from a different part of the docs - i don't want to update anything in `email.rst` for this PR

File: mailer.rst (hunk 16 above)
Comment (weaverryan) [reply mailer.rst]: good catch!

File: mailer.rst (hunk 17 above)
Comment (weaverryan) [reply mailer.rst]: I did it alphabetically, but put gmail last, as that's not really a great option

File: mailer.rst (hunk 18 above)
Comment (weaverryan) [reply mailer.rst]: ha! nice catch - a bad copy from some local code when I was building all of this!

**Comment by weaverryan** [comment]
Thanks to everyone for the wonderful review - a lot of improvements and fixes came because of that!

**Pull request merged**